### ISSUE TITLE
Clarification on Predicate Semantics

### ISSUE THREAD

**Issue by xogeny** [jsonata-js/jsonata#154]
I think the predicate semantics need clarifying.  In the discussion on [basic selection](http://docs.jsonata.org/basic.html#navigating-json-arrays), it says:

> If the square brackets contains a number, or an expression that evaluates to a number, then the number represents the index of the value to select... If the expression in square brackets is non-numeric, or is an expression that doesn’t evaluate to a number, then it is treated as a predicate.

Under [predicates](http://docs.jsonata.org/complex.html#predicates) (the link for predicate in the above quote is wrong, BTW), it says:

> the selected items can be filtered using a predicate - [expr] where expr evaluates to a Boolean value. Each item in the selection is tested against the expression, if it evaluates to true, then the item is kept; if false, it is removed from the selection.

I think the case where the expression evaluates to a scalar number is clear.  But I do see a couple of problems.  First, the first quote says (basically), if the expression doesn't evaluate to a number, treat it as a predicate.  Any array is not a number, so the behavior of arrays should be described under predicates...but it isn't.  It describes the case where `$` is an array (apply predicate to each element in that array) but not what happens if the expression is an array.  It only describes the semantics of the predicate expression is a boolean.

I see two ways to handle a predicate expression that evaluates to an array.  But I think the only one that really makes sense is the one currently in the code which basically decides that if the array is an array of numbers, then use the "basic" evaluation but apply it element-wise to `$` to construct an array (if I understand the current implementation correctly).

So I think for this issue it is mainly an issue of documentation (arrays just aren't discussed...at least not in the places I mentioned).  But the other issue I see is an inconsistency between the documentation and the implementation.  For predicates, it says "where expr evaluates to a Boolean".  But this isn't actually what is done in the code.  [What actually happens](https://github.com/jsonata-js/jsonata/blob/master/jsonata.js#L4165-L4172) is that it looks for "truthy" values.  So these things don't need to actually evaluate to Booleans, but pretty much any Javascript value.  But that's where it gets confusing, because empty arrays are `false` and non-empty arrays are `true` using "truthy" evaluation.  Consider:

Input: `[0,1,2]`
Expr: `$[[0,1]]`
Result: `[0,1]`

According to the current documentation, the predicate expression is not numeric.  As such, it should be a predicate.  Applying truthy semantics, this expression should just be `true` and return everything.  But a special case is carved out (in the code, but not in the documentation as far as I could tell) for arrays of integers and how those are handled.  This is intuitive.  So I don't have issue with the semantics as much as the documentation.  But, as with all special cases, you get some odd corner cases, *e.g.,*

Input: `[0,1,2]`
Expr: `$[[0,1,"foo"]]`
Result: `[0,1,2]`

This is a bit counter-intuitive.  What is happening here is that the expression isn't an array of numbers so it is a "predicate" and since it is a non-empty array it is `true`.  There are two issues.  The first is the potential confusion cause by:

Input: `[0,1,2]`
Expr: `$[[0,1,"0"]]`
Result: `[0,1,2]`

Someone might expect to get back `[0,1,0]` since, in Javascript, you can use the string `"0"` to dereference an array.  So we are using some Javascript semantics (truthy evaluation) but not others.  So this needs to be factored into the documentation.  Note that the filtering of `undefined` values from arrays is **essential** to explain here because something like:

Input: `[0,1,2]`
Expr: `$[[0,1,$.foo]]`
Result: `[0,1]`

That expression only evaluates to an array of integer (`[0, 1]`) because `$.foo` evaluates to `undefined` and is, as such, *dropped* from the result (*i.e.,* `[0,1,$.foo]` evaluates to `[0,1]` and not `[0,1,undefined]` because the latter is not an array of integers).

In summary, the explanation of array handling is essentially in the documentation since it really represents a special case.

**Comment by jianatldd** [comment]
I think this is not a problem just need clarification. the root of the issue is the context the predicate expression evaluate against, Guess if the expression is intended to be an index or condition by check type of result cannot solve the context issue. 

in the old xpath days. there is no index for an array option ( have to use position()=1).  so it is clear the expression inside [] is evaluated for each item of the array. But no one will understand why an index of array expression is actually evaluated against the item of the array. 

example: (in test page)
Account.Order[Account.LastOrder]  
return nothing when set LastOrder in Account to 0, 1 
this one
Account.Order[Product[1].Quantity]  
return 2nd Order because  Product[1].Quantity in first order is set to 1,
but everyone expects it should return both Orders because it looks like predicate expression, not an index

I think we need a prefix to indicate what is the purpose of the expression inside []

**Comment by jianatldd** [comment]
another way is to be very restrictive. the expression has to be clearly a boolean type ( it has to be a result of a boolean operator, even in the case the value is a boolean,  i.e. Account.Order[Complete= true] is good but Account.Order[Complete] is not good, it never makes sense for an index to be a boolean 

The compiler can detect expression return type and evaluate in the right context

**Comment by xogeny** [comment]
As I said, I don't see this as an issue with the semantics, just with the documentation of the semantics.  It is merely the omission of the array handling that was an issue. I've re-implemented this myself and I managed to do it in only about 10 lines of code (leverage some high-level utility functions).  It all hinges on figuring out that you evaluate the predicate expression for each possible "lhs" value and if it is purely an array of numbers, then you look for the index of the current lhs value in that array and, if present, you keep the value.  Otherwise, if the predicate isn't an array of numbers, just treat it with truthy semantics.  Wasn't so bad once I figured it out.  But I had to read the code (for a while) to really understand what was going on.  This should all be reflected in the documentation somewhere was my only point.

**Comment by jianatldd** [comment]
I think the issue you have raised is not same as mine, maybe mine should be in a new thread, the problem i have is not how to match the result of predicate expression, but in which context the expression is evaluated. 

for an index value of an array, it is natural for a javascript programmer to think that it would be evaluated in the same context as the left side expression  (i.e. Account.Order[ Account.LastOrderNumber ] ), but for a predicate expression, it is evaluated in the context of every item in the array, the current implementation here is to evaluate for every item in the context of the item. (natural for anyone from the xpath, xslt background)

this ambiguity makes the use of []  dangerous, As is now, to avoid confusion,  the index of array should only be used when expression inside [] only contains constants and global  ($$) variables and result to a number (not even string like this "1" because Account.Order['0'] actually return all Orders because "0" is true for all items )

**Comment by andrew-coleman** [comment]
I added support for filtering based on an array of indexes in this [commit](https://github.com/jsonata-js/jsonata/commit/86bde0859210e4f106be9238c8d5351708d42b04) and clearly forgot to document it - I'll fix that!

If the filter expression evaluates to an array of numbers, then those numbers represent the items within the sequence being tested to be kept.  For example, 
`['a','b','c','d','e','f','g','h'][[0..2,4,-1]]` evaluates to `[ "a", "b", "c", "e", "h"]`
because it selects indexes `0` to `2`, `4`, and the last item (negative indexes count from the end).

I think this shows that some time and effort needs to be applied to the documentation in general.  A few issues are being created that boil down to insufficient docs (e.g. [here](https://github.com/jsonata-js/jsonata/issues/143#issuecomment-362951196) and [here](https://github.com/jsonata-js/jsonata/issues/157)).  We need to make sure that these are captured as issues in the [docs repo](https://github.com/jsonata-js/jsonata-js.github.io) (the documentation is also open-source).  I also invite anyone to contribute improvements to the documentation for this project.  This can come in many forms, e.g. direct edits/additions to existing docs, new tutorials/how-to guides, blog posts that I can link to, etc.  @shrickus, you have been very supportive and evangelized the merits of jsonata in the Node-RED community - do you have any suggestions relating to this?

Thanks!

**Comment by xogeny** [comment]
Yes, these evaluation semantics confused me at first.  Initially, I thought the semantics would be to evaluate the predicate and then pluck out the left hand side values according to the numbers that the predicate resolved to.  But then this was at odds with the element-wise nature you need for logical predicates.  So then I figured out that what I really needed to was do the evaluation element-wise (*i.e.,* evaluate the predicate in the context of every element on the LHS) and then check to see if the predicate evaluated to an array of numbers and, if so, that the index of the current LHS value being examined was included among the indices returned by the predicate.  It makes sense once you consider how the logical stuff has to work and it seems to largely impact the implementation, not the user.  In any case, I did it the same way and my evaluator seems to be passing the test cases there.
